ckan/views/resource.py

```diff
@@ -0,0 +1,908 @@
+            to_preview=False,
+            pkg_dict=pkg_dict,
+            pkg=pkg,
+            resource=resource
+        )
+        return context, extra_vars
+
+    def post(self, package_type, id, resource_id, view_id=None):
+        context, extra_vars = self._prepare(id, resource_id)
+
+        to_preview = request.POST.pop(u'preview', False)
+        if to_preview:
+            context[u'preview'] = True
+        to_delete = request.POST.pop(u'delete', None)
+        data = clean_dict(
+            dict_fns.unflatten(
+                tuplize_dict(
+                    parse_params(request.form, ignore_keys=CACHE_PARAMETERS)
+                )
+            )
+        )
+        data.pop(u'save', None)
+        data[u'resource_id'] = resource_id
+
+        try:
+            if to_delete:
+                data[u'id'] = view_id
+                get_action(u'resource_view_delete')(context, data)
+            elif view_id:
+                data[u'id'] = view_id
+                data = get_action(u'resource_view_update')(context, data)
+            else:
+                data = get_action(u'resource_view_create')(context, data)
+        except ValidationError as e:
+            # Could break preview if validation error
+            to_preview = False
+            extra_vars[u'errors'] = e.error_dict,
+            extra_vars[u'error_summary'] = e.error_summary
+        except NotAuthorized:
+            # This should never happen unless the user maliciously changed
+            # the resource_id in the url.
+            return base.abort(403, _(u'Unauthorized to edit resource'))
+        else:
+            if not to_preview:
+                return h.redirect_to(
+                    u'resource.views', id=id, resource_id=resource_id
+                )
+        extra_vars[u'data'] = data
+        extra_vars[u'to_preview'] = to_preview
+        return self.get(package_type, id, resource_id, view_id, extra_vars)
+
+    def get(
+        self, package_type, id, resource_id, view_id=None, post_extra=None
+    ):
+        context, extra_vars = self._prepare(id, resource_id)
+        to_preview = extra_vars[u'to_preview']
+        if post_extra:
+            extra_vars.update(post_extra)
+
+        package_type = _get_package_type(id)
+
+        # view_id exists only when updating
+        if view_id:
+            try:
+                old_data = get_action(u'resource_view_show')(
+                    context, {
+                        u'id': view_id
+                    }
+                )
+                data = extra_vars[u'data'] or old_data
+                view_type = old_data.get(u'view_type')
+                # might as well preview when loading good existing view
+                if not extra_vars[u'errors']:
+                    to_preview = True
+            except (NotFound, NotAuthorized):
+                return base.abort(404, _(u'View not found'))
+
+        view_type = view_type or request.args.get(u'view_type')
```

Comment:
@smotornyuk here im getting error `UnboundLocalError: local variable 'view_type' referenced before assignment
` when clicking on 'New view' button when editing views of the resource